demos/custom-tabs-example-app/src/main/java/org/chromium/customtabsdemos/EngagementSignalsActivity.java

```diff
@@ -0,0 +1,193 @@
+// Copyright 2023 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package org.chromium.customtabsdemos;
+
+import static androidx.browser.customtabs.CustomTabsIntent.ACTIVITY_HEIGHT_ADJUSTABLE;
+
+import android.net.Uri;
+import android.os.Bundle;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.appcompat.app.AppCompatActivity;
+
+import android.os.RemoteException;
+import android.util.Log;
+import android.view.View;
+import android.widget.EditText;
+import android.widget.TextView;
+
+import androidx.browser.customtabs.CustomTabsCallback;
+import androidx.browser.customtabs.CustomTabsClient;
+import androidx.browser.customtabs.CustomTabsIntent;
+import androidx.browser.customtabs.CustomTabsSession;
+import androidx.browser.customtabs.EngagementSignalsCallback;
+
+public class EngagementSignalsActivity extends AppCompatActivity implements View.OnClickListener {
+    private static final String TAG = EngagementSignalsActivity.class.getSimpleName();
+    private static final int INITIAL_HEIGHT_DEFAULT_PX = 600;
+
+    private EditText mUrlEditText;
+    private TextView mTextVerticalScroll;
+    private TextView mTextGreatestPercentage;
+    private TextView mTextSessionEnd;
+    private TextView mTextNavigation;
+
+    private ServiceConnection mConnection;
```

Comment:
Could you add `@Nullable` to these variables when appropriate?

(We can probably leave the `EditText`/`TextView`s without the annotation as they're practically non-nullable after onCreate has been called. It's not strictly correct, but I think is reasonable.)

Reply:
👍